test/widgets/lightbox_test.dart

```diff
@@ -197,6 +203,136 @@ class FakeVideoPlayerPlatform extends Fake
+    final message = eg.streamMessage(stream: channel,
+      contentMarkdown: ContentExample.imageSingle.html, topic: 'test topic');
+
+    Future<void> setupMessageListPage(WidgetTester tester) async {
+      addTearDown(testBinding.reset);
+      final subscription = eg.subscription(channel);
+      await testBinding.globalStore.add(eg.selfAccount, eg.initialSnapshot(
+        streams: [channel], subscriptions: [subscription]));
+      store = await testBinding.globalStore.perAccount(eg.selfAccount.id);
+      connection = store.connection as FakeApiConnection;
+      await store.addUser(eg.selfUser);
+
+      connection.prepare(json:
+        eg.newestGetMessagesResult(foundOldest: true, messages: [message]).toJson());
+
+      await tester.pumpWidget(TestZulipApp(accountId: eg.selfAccount.id,
+        child: MessageListPage(initNarrow: const CombinedFeedNarrow())));
+
+      await tester.pumpAndSettle();
+    }
+
+    testWidgets('Hero animation occurs smoothly when opening lightbox from message list', (tester) async {
+      prepareBoringImageHttpClient();
+
+      await setupMessageListPage(tester);
+
+      final messageContentFinder = find.byWidgetPredicate((widget) =>
+        widget is MessageContent && widget.message.id == message.id
+      );
+      final messageListImageFinder = find.descendant(
+        of: messageContentFinder,
+        matching: find.byType(RealmContentNetworkImage)
+      );
+      final imageSrcUrl = tester
+        .widget<RealmContentNetworkImage>(messageListImageFinder).src;
+      final initialImagePosition = tester.getRect(messageListImageFinder);
+      await tester.tap(messageListImageFinder);
+      await tester.pump();
+      // pump to start hero animation
+      await tester.pump();
+
+      const heroAnimationDuration = Duration(milliseconds: 300);
+      const steps = 150;
+      final stepDuration = heroAnimationDuration ~/ steps;
+      List<Rect> animatedPositions = [];
+      for (int i = 1; i <= steps; i++) {
+        await tester.pump(stepDuration);
+
+        final animatedFlightImageFinder = find.byWidgetPredicate((widget) =>
+          widget is RealmContentNetworkImage && widget.src == imageSrcUrl
+        );
+        final animatedFlightImagePosition = tester.getRect(animatedFlightImageFinder);
+        animatedPositions.add(animatedFlightImagePosition);
+      }
+
+      final totalTopDistance = (
+        initialImagePosition.top - animatedPositions.last.top).abs();
+      final totalLeftDistance = (
+        initialImagePosition.left - animatedPositions.last.left).abs();
+
+      Rect previousPosition = initialImagePosition;
+      for (final position in animatedPositions) {
+        final positionChangeTop = position.top - previousPosition.top;
+        final positionChangeLeft = position.left - previousPosition.left;
+        check(positionChangeTop.abs()).isLessThan(0.1 * totalTopDistance);
+        check(positionChangeLeft.abs()).isLessThan(0.1 * totalLeftDistance);
```

Comment:
10% of the distance traveled in the whole animation is quite a lot for a duration of 1/150th of the total duration.

We can measure this better by using the Pythagorean theorem to compute the overall distance moved in the two-dimensional plane, rather than looking at motion in each axis separately. That way if the start and end points happen to be close in one axis, this doesn't expect unreasonably tight bounds on how much the image might move in that axis during the animation.